### PR TITLE
A first effort in tackling the warnings described in issue 64

### PR DESCRIPTION
**Pull request by NiMlr**
See first and second entry in [issue 64](https://github.com/mdp-toolkit/mdp-toolkit/issues/61).

### CODE REVIEW

**Comment by NiMlr** [comment]
The second fix is copied from [this code](https://github.com/nltk/nltk/blob/81490305bbbb42c6b7fa62a4694d65607b7f420f/nltk/decorators.py#L28). I am not sure which part of MDP is best to include the legacy adaption code, as it is needed for all occurrences of `formatargspec`. Note, that I have not fixed all of these occurrences yet.

File: mdp/utils/routines.py [hunk 1]
```diff
@@ -11,6 +11,22 @@
...
+    Basically, from a Signature we create a string and remove the default values.
+    """
+    listsignature = str(signature)[1:-1].split(",")
+    for counter, param in enumerate(listsignature):
+        if param.count("=") > 0:
+            listsignature[counter] = param[0:param.index("=")].strip()
+        else:
+            listsignature[counter] = param.strip()
+    return ", ".join(listsignature)
+
```
Comment (Stewori) [inline mdp/utils/routines.py]: I think on top-level PEP 8 suggests two lines to separate functions. I've seen this pattern frequently in MDP so far.

File: mdp/utils/routines.py [hunk 2]
```diff
@@ -11,6 +11,22 @@
 import random
 import itertools
 
+def _legacysignature(signature):
+    """
+    Used to replace inspect.formatargspec by inspect.signature.
```
Comment (Stewori) [inline mdp/utils/routines.py]: ->    `"""Used to replace inspect.formatargspec `

File: mdp/signal_node.py [hunk 3]
```diff
@@ -187,7 +174,14 @@ def _function_infodict(func):
             argnames.append(varargs)
         if varkwargs:
             argnames.append(varkwargs)
-        signature = _legacysignature(inspect.signature(func))
+        try:
+            signature = _legacysignature(inspect.signature(func))
+        except:
```
Comment (Stewori) [inline mdp/signal_node.py]: Do we know the raised exception more specifically? I think a plain `except` should be avoided if reasonably possible.

**Comment by Stewori** [comment]
Difficult to tell where it should go.
I see these possibilities: configuration.py, helper_funcs.py, utils/introspection.py or make a custom file, e.g. utils/compatibility.py
Maybe helper_funcs would be best?

**Comment by Stewori** [comment]
And thanks for this work. I'd be happy to include it into the upcoming release.

**Comment by NiMlr** [comment]
> Difficult to tell where it should go.
> I see these possibilities: configuration.py, helper_funcs.py, utils/introspection.py or make a custom file, e.g. utils/compatibility.py
> Maybe helper_funcs would be best?

Based from the contents of the files, I think it fits better in routine.py than in helper_funcs.py.

File: mdp/signal_node.py (hunk 3 above)
Comment (NiMlr) [reply mdp/signal_node.py]: Yes, you are right. It should be an `AttributeError`. See [build](https://travis-ci.org/mdp-toolkit/mdp-toolkit/jobs/625163939#L4622).

**Comment by Stewori** [comment]
It appeared to me that routines mostly consists of data-processing routines. However, the same is true for helper_funcs but the name sounded more generic to me. Anyway, I don't really care. routines is fine. Please separate all top-level functions in routines by two blank lines if you put it there. Thanks.

**Comment by NiMlr** [comment]
> It appeared to me that routines mostly consists of data-processing routines. However, the same is true for helper_funcs but the name sounded more generic to me. Anyway, I don't really care. routines is fine. Please separate all top-level functions in routines by two blank lines if you put it there. Thanks.

Should be included in the most recent commit.

File: mdp/signal_node.py [hunk 4]
```diff
@@ -173,10 +174,15 @@ def _function_infodict(func):
...
-                                          defaults,
+
+        # python 2/3 compatibility
+        try:
+            signature = _legacysignature(inspect.signature(func))
+        except AttributeError:
+            signature = inspect.formatargspec(regargs,	      
+                                          varargs,	
+                                          varkwargs,	
+                                          defaults,
```
Comment (Stewori) [inline mdp/signal_node.py]: Either fix the alignment or put it like this:
```
signature = inspect.formatargspec(
        regargs, varargs, varkwargs, defaults, formatvalue=lambda value: "")[1:-1]
```
(additional linebreak as needed)

Comment (NiMlr) [reply mdp/signal_node.py]: The alignment should not be different, as then the column limit is exceeded by the last argument.

Comment (Stewori) [reply mdp/signal_node.py]: Please make it a three-liner then. This non-alignment just looks wrong.
```
signature = inspect.formatargspec(
        regargs, varargs, varkwargs, defaults,
        formatvalue=lambda value: "")[1:-1]
```

**Comment by Stewori** [comment]
I will merge a traivs.yml update first because that will save 7-10min for future travis runs.

**Comment by NiMlr** [comment]
I still have to fix the other occurrences of `fortmatargspec` anyway.

**Comment by Stewori** [comment]
Please also add a line to CHANGES on that occasion:
`2019-15-12: FIX: Warnings type interpretation and formatargspec deprecation`

**Comment by NiMlr** [comment]
There is a problem. The implementation of `_legacysignature` is not general enough in transforming the output of `inspect.signature` to the one of `inspect.formatargspec`. In order to only use the new `inspect.signature` we would need to rewrite whatever functions use it downstream - so I tried to avoid it. Others have avoided it by [implementing the functionality of `inspect.formatargspec`](https://github.com/NikolayPogoreliy/sqla_tests/commit/536d5187a038a44aec624dd2a99792f49dec82ed). I am trying to do this now.

File: mdp/signal_node.py [hunk 5]
```diff
@@ -175,15 +176,10 @@ def _function_infodict(func):
...
-        try:
-            signature = _legacysignature(inspect.signature(func))
-        except AttributeError:
-            signature = inspect.formatargspec(regargs,	      
-                                          varargs,	
-                                          varkwargs,	
-                                          defaults,	
-                                          formatvalue=lambda value: "")[1:-1]
+        signature = inspect_formatargspec(
+                regargs,	varargs, varkwargs,	defaults,
```
[the remark this answers is not in the document]
Comment (NiMlr) [reply mdp/signal_node.py]: Fixed it.

**Comment by NiMlr** [comment]
The PR can merged if you are ok with it.

**Comment by Stewori** [comment]
Thanks again for the good solution.

**Pull request merged**